lib/improver/metadata/forecast_times.py

```diff
@@ -37,12 +37,14 @@
 import iris
 from iris.exceptions import CoordinateNotFoundError
 
+from improver.metadata.constants.time_types import (
+    TIME_INTERVAL_DTYPE, TIME_INTERVAL_UNIT, TIME_REFERENCE_UNIT)
 from improver.utilities.cube_manipulation import build_coordinate
 from improver.utilities.temporal import cycletime_to_datetime
 
 
-def forecast_period_coord(
-        cube, force_lead_time_calculation=False, result_units="seconds"):
+def forecast_period_coord(cube, force_lead_time_calculation=False,
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Too many branches (13/12)](https://app.codacy.com/app/metoppv_tech/improver/pullRequest?prid=4600460)

Reply:
Here is a quick attempt at lifting the function out.
Would need to name the function.

```
def forecast_period_coord(
        cube, force_lead_time_calculation=False, result_units="seconds"):
    """
    Return or calculate the lead time coordinate (forecast_period)
    within a cube, either by reading the forecast_period coordinate,
    or by calculating the difference between the time (points and bounds) and
    the forecast_reference_time. The units of the forecast_period, time and
    forecast_reference_time coordinates are converted, if required. The final
    coordinate will have units of seconds.

    Args:
        cube (iris.cube.Cube):
            Cube from which the lead times will be determined.
        force_lead_time_calculation (bool):
            Force the lead time to be calculated from the
            forecast_reference_time and the time coordinate, even if
            the forecast_period coordinate exists.
            Default is False.
        result_units (str or cf_units.Unit):
            Desired units for the resulting forecast period coordinate.

    Returns:
        iris.coords.Coord:
            Describing the points and their units for
            'forecast_period'. A DimCoord is returned if the
            forecast_period coord is already present in the cube as a
            DimCoord and this coord does not need changing, otherwise
            it will be an AuxCoord. Units are result_units.
    """
    fp_type = (cube.coord("forecast_period").dtype if 
               cube.coords("forecast_period") else np.int32)

    if cube.coords("forecast_period") and not force_lead_time_calculation:
        result_coord = cube.coord("forecast_period").copy()
        try:
            result_coord.convert_units(result_units)
        except ValueError as err:
            msg = "For forecast_period: {}".format(err)
            raise ValueError(msg)

    # Try to return forecast_reference_time - time coordinate.
    elif cube.coords("time") and cube.coords("forecast_reference_time"):
        result_coord = method_name(cube, result_units)
    else:
        msg = ("The forecast period coordinate is not available within {}."
               "The time coordinate and forecast_reference_time "
               "coordinate were also not available for calculating "
               "the forecast_period.".format(cube))
        raise CoordinateNotFoundError(msg)

    result_coord.points = result_coord.points.astype(fp_type)
    if result_coord.bounds is not None:
        result_coord.bounds = result_coord.bounds.astype(fp_type)

    return result_coord


def method_name(cube, result_units):
    time_units = cube.coord("time").units
    t_coord = cube.coord("time")
    fr_coord = cube.coord("forecast_reference_time")
    fr_type = fr_coord.dtype
    try:
        fr_coord.convert_units(time_units)
    except ValueError as err:
        msg = "For forecast_reference_time: {}".format(err)
        raise ValueError(msg)
    time_points = np.array([c.point for c in t_coord.cells()])
    forecast_reference_time_points = np.array(
        [c.point for c in fr_coord.cells()])
    required_lead_times = (time_points - forecast_reference_time_points)
    required_lead_times = np.array(
        [x.total_seconds() for x in required_lead_times]).astype(fr_type)
    if t_coord.bounds is not None:
        time_bounds = np.array([c.bound for c in t_coord.cells()])
        required_lead_bounds = (
                time_bounds - forecast_reference_time_points)
        required_lead_bounds = np.array(
            [[b.total_seconds() for b in x]
             for x in required_lead_bounds]).astype(fr_type)
    else:
        required_lead_bounds = None
    coord_type = iris.coords.AuxCoord
    if cube.coords("forecast_period"):
        if isinstance(
                cube.coord("forecast_period"), iris.coords.DimCoord):
            coord_type = iris.coords.DimCoord
    result_coord = coord_type(
        required_lead_times,
        standard_name='forecast_period',
        bounds=required_lead_bounds,
        units="seconds")
    result_coord.convert_units(result_units)
    if np.any(result_coord.points < 0):
        msg = ("The values for the time {} and "
               "forecast_reference_time {} coordinates from the "
               "input cube have produced negative values for the "
               "forecast_period. A forecast does not generate "
               "values in the past.").format(
            cube.coord("time").points,
            cube.coord("forecast_reference_time").points)
        warnings.warn(msg)
    return result_coord
```

Reply:
Here is my quick fix for this.
The function would require renaming...
...and docstrings 

```python
def forecast_period_coord(cube, force_lead_time_calculation=False,
                          result_units=TIME_INTERVAL_UNIT):
    """
    Return or calculate the lead time coordinate (forecast_period)
    within a cube, either by reading the forecast_period coordinate,
    or by calculating the difference between the time (points and bounds) and
    the forecast_reference_time. The units of the forecast_period, time and
    forecast_reference_time coordinates are converted, if required. The final
    coordinate will have units of seconds.

    Args:
        cube (iris.cube.Cube):
            Cube from which the lead times will be determined.
        force_lead_time_calculation (bool):
            Force the lead time to be calculated from the
            forecast_reference_time and the time coordinate, even if
            the forecast_period coordinate exists.
            Default is False.
        result_units (str or cf_units.Unit):
            Desired units for the resulting forecast period coordinate.

    Returns:
        iris.coords.Coord:
            Describing the points and their units for
            'forecast_period'. A DimCoord is returned if the
            forecast_period coord is already present in the cube as a
            DimCoord and this coord does not need changing, otherwise
            it will be an AuxCoord. Units are result_units.
    """
    if cube.coords("forecast_period"):
        fp_type = cube.coord("forecast_period").dtype
    else:
        fp_type = TIME_INTERVAL_DTYPE

    if cube.coords("forecast_period") and not force_lead_time_calculation:
        result_coord = cube.coord("forecast_period").copy()
        try:
            result_coord.convert_units(result_units)
        except ValueError as err:
            msg = "For forecast_period: {}".format(err)
            raise ValueError(msg)

    # Try to return forecast_reference_time - time coordinate.
    elif cube.coords("time") and cube.coords("forecast_reference_time"):
        result_coord = method_name(cube, result_units)
    else:
        msg = ("The forecast period coordinate is not available within {}."
               "The time coordinate and forecast_reference_time "
               "coordinate were also not available for calculating "
               "the forecast_period.".format(cube))
        raise CoordinateNotFoundError(msg)

    result_coord.points = result_coord.points.astype(fp_type)
    if result_coord.bounds is not None:
        result_coord.bounds = result_coord.bounds.astype(fp_type)

    return result_coord


def method_name(cube, result_units):
    time_units = cube.coord("time").units
    t_coord = cube.coord("time")
    fr_coord = cube.coord("forecast_reference_time")
    fr_type = fr_coord.dtype
    try:
        fr_coord.convert_units(time_units)
    except ValueError as err:
        msg = "For forecast_reference_time: {}".format(err)
        raise ValueError(msg)
    time_points = np.array(
        [c.point for c in t_coord.cells()])
    forecast_reference_time_points = np.array(
        [c.point for c in fr_coord.cells()])
    required_lead_times = (
            time_points - forecast_reference_time_points)
    required_lead_times = np.array(
        [x.total_seconds() for x in required_lead_times]).astype(fr_type)
    if t_coord.bounds is not None:
        time_bounds = np.array(
            [c.bound for c in t_coord.cells()])
        required_lead_bounds = (
                time_bounds - forecast_reference_time_points)
        required_lead_bounds = np.array(
            [[b.total_seconds() for b in x]
             for x in required_lead_bounds]).astype(fr_type)
    else:
        required_lead_bounds = None
    coord_type = iris.coords.AuxCoord
    if cube.coords("forecast_period"):
        if isinstance(
                cube.coord("forecast_period"), iris.coords.DimCoord):
            coord_type = iris.coords.DimCoord
    result_coord = coord_type(
        required_lead_times,
        standard_name='forecast_period',
        bounds=required_lead_bounds,
        units="seconds")
    result_coord.convert_units(result_units)
    if np.any(result_coord.points < 0):
        msg = ("The values for the time {} and "
               "forecast_reference_time {} coordinates from the "
               "input cube have produced negative values for the "
               "forecast_period. A forecast does not generate "
               "values in the past.").format(
            cube.coord("time").points,
            cube.coord("forecast_reference_time").points)
        warnings.warn(msg)
    return result_coord
```